pkg/cluster/clusterserviceprincipal.go

```diff
@@ -0,0 +1,163 @@
+package cluster
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"strings"
+
+	"github.com/ghodss/yaml"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+
+	"github.com/Azure/ARO-RP/pkg/util/arm"
+	"github.com/Azure/ARO-RP/pkg/util/rbac"
+	"github.com/Azure/ARO-RP/pkg/util/stringutils"
+	"github.com/Azure/azure-sdk-for-go/services/preview/authorization/mgmt/2018-09-01-preview/authorization"
+)
+
+func (m *manager) createOrUpdateClusterServicePrincipalRBAC(ctx context.Context) error {
+	resourceGroupID := m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID
+	resourceGroup := stringutils.LastTokenByte(resourceGroupID, '/')
+
+	clusterSPObjectID, err := m.aad.GetServicePrincipalID(ctx)
+	if err != nil {
+		return err
+	}
+
+	roleAssignments, err := m.roleAssignments.ListForResourceGroup(ctx, resourceGroup, "")
+	if err != nil {
+		return err
+	}
+
+	// If we have Contributor RBAC role for Cluster SP on the resource group in question
+	// We are interested in Resource group scope only (inherited are returned too).
+	var toDelete []authorization.RoleAssignment
+	var found bool
+	for _, assignment := range roleAssignments {
+		// Contributor assignments only!
+		if *assignment.Scope == resourceGroupID && strings.HasSuffix(*assignment.RoleDefinitionID, rbac.RoleContributor) {
+			if *assignment.PrincipalID == clusterSPObjectID {
+				found = true
+			} else {
+				toDelete = append(toDelete, assignment)
+			}
+		}
+	}
+
+	if len(toDelete) > 0 {
+		for _, assignment := range toDelete {
+			m.log.Infof("Deleting Contributor roleAssignment %s", *assignment.Name)
+			_, err := m.roleAssignments.Delete(ctx, *assignment.Scope, *assignment.Name)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if !found {
+		m.log.Info("Contributor roleAssignment not found for cluster service principal. Creating")
+		t := &arm.Template{
+			Schema:         "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+			ContentVersion: "1.0.0.0",
+			Resources:      []*arm.Resource{m.clusterServicePrincipalRBAC(clusterSPObjectID)},
+		}
+
+		err = m.arm.Deploy(ctx, resourceGroup, "storage", t, nil)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
@jim-minter, this is logic which causes me to worry.
We use `"guid(resourceGroup().id, 'SP / Contributor')"` for name. So we need to DELETE first before we try to create. (Update is not possible in AAD).
This makes delete and create bit tricky. But I think we will have to accept this.

In addition this step will have to be first in Credentials update flow, otherwise it might get us into limbo state.